prov/rxm/src/rxm_ep.c

```diff
@@ -730,6 +730,31 @@ rxm_multi_recv_entry_get(struct rxm_ep *rxm_ep, const struct iovec *iov,
 	return recv_entry;
 }
 
+struct rxm_tx_buf *rxm_get_tx_buf(struct rxm_ep *ep)
+{
+	struct rxm_tx_buf *buf;
+
+	assert(fastlock_held(&ep->util_ep.lock));
+	if (!ep->tx_credit)
+		return NULL;
```

Comment:
@shefty - This impacts the allocating a TX buffer for the atomic response. Previously we would allow the response to get allocated and later be deferred for TX if necessary; we were limiting the atomic request based on atomic specific TX credits. I believe now it will fail at allocation in CQ processing of the request. One case that could problematic is an application that loops back to itself as to access atomic updated data via consistent local/remote flow. If you agree this could be a problem, then one option might be to defer processing of the atomic request in the CQ processing if a TX buffer for the atomic response is not available.

Reply:
The intent is that only user initiated transfers need to acquire a credit.  All internally generated transfers bypass the credit check and will only fail if the system is out of memory.  A prior changed removed the caps on the buffer pools, allowing them to grow.  The code is not structured to handle failures when it needs to send an internal message.  Internal allocations still call ofi_buf_alloc() directly, rather than this abstraction.

Reply:
Although, now that I think about it, deferring the atomic request before processing the atomic operation could end up in out of order processing if another atomic comes in before the deferred atomic is processed; so not a good option. I think that was the driving force in letting the atomic response always acquire a TX buffer.

Reply:
> The intent is that only user initiated transfers need to acquire a credit. All internally generated transfers bypass the credit check and will only fail if the system is out of memory. A prior changed removed the caps on the buffer pools, allowing them to grow. The code is not structured to handle failures when it needs to send an internal message. Internal allocations still call ofi_buf_alloc() directly, rather than this abstraction.

Makes sense, thanks for the clarification.

Reply:
It looks like we're in agreement.

The problem is greater than atomics.  We're basically in a receiver side overrun, impacting sending any ack, completing rendezvous transfers, etc.  The code cannot handle a failure through these paths, but should now only occur when we're out of memory.  When we hit that case, we basically need to send the endpoint into an error state.

The old, old code set caps on the size of the internal pools, which made it more likely that we would fail to allocate a message for an internal transfer.  That wasn't an out of memory problem, but still results in the transfers never completing.  I merged all the separate pools together to reduce the memory footprint, but removed the pool cap, so internal transfers wouldn't be limited.  This change adds a cap back for user initiated transfers (one we can return EAGAIN to the user).

In order to handle overrun errors, we need some sort of changes at the receiver.  For example, if the receiver cannot allocate a response resources, it could re-use the receive buffer to send some sort of retry request to the initiator.  That would require significant code restructuring.  Or we allocate a tx buffer with each rx buffer, so that a response message is always available.  But that increases the memory footprint.

Anyway, this is my happy compromise, because we're seeing failures resulting from being out of memory.

Reply:
Yes, I agree. That was the intent of the atomic implementation, and making it generic for all software protocols is a good compromise.